### PR TITLE
feat(baselines) Add baseline FedBABU

### PR DESCRIPTION
**Pull request by KarhouTam**
## Issue
Add new baseline FedBABU
https://arxiv.org/abs/2106.06042

### Related issues/PRs

Placeholder issue: #4285

### CODE REVIEW

**Comment by KarhouTam** [comment]
Hi, @jafermarq 
I found that the deps versions in baseline template are somehow outdated. 

For example, I need `flwr_datasets.preprocessor.Merger`, but `flwr_datasets 0.1.0` will be installed by `pip install -e .` command and it doesn't support that. Maybe the deps verisions need to be updated.

**Comment by jafermarq** [comment]
Hi @KarhouTam, what version of flower I you using? The flower baselines template is using the latest release as a minimum version: https://github.com/adap/flower/blob/b8574258d32bab50553ceada45fd5251a9e5e7be/src/py/flwr/cli/new/templates/app/pyproject.baseline.toml.tpl#L12

Note to create the baseline, it's best to do it like this (since it will also populate a `README.md` and create auxiliary files you might need -- it's very similar to the previous baseline format but now using `flwr new/run` style):

```
flwr new fedbabu 
# then choose Flower Baseline template (should be option 8)
```

**Comment by KarhouTam** [comment]
Seems I forgot to update my local flwr version first 😂. Nevermind.

**Comment by WilliamLindskog** [comment]
Hi @KarhouTam,

Just checking in here. Seems like the implementation is nearly passing all tests. Are there any issues you are facing?

**Comment by KarhouTam** [comment]
Thanks, @WilliamLindskog .

Everything is fine so far. I am currently busy on other things so this PR is panding for now. Once I am free, the process will be re-launched. I will ask for help if I meet problems in that time.

**Comment by WilliamLindskog** [comment]
Hi @KarhouTam, 

Just checking in here to see if there's anything we can support you with.

**Comment by WilliamLindskog** [comment]
Sure thing, I might close related issue if this is already being worked on and migrate this baseline to Flower Discuss if someone wants to pick this up before we can merge it.

**Comment by WilliamLindskog** [comment]
@jafermarq seems like most tests pass but some baseline stuff doesn't run, have you seen this before? One error doesn't really show an output log. @KarhouTam did you encounter an error when running locally?

**Comment by KarhouTam** [comment]
Actually the algorithm code is not finished yet. I'm pushing the process locally and will update this thread when most things are done.

**Comment by KarhouTam** [comment]
Hi, @jafermarq @WilliamLindskog . I have some questions within the reproduction.
1. How to get the number of current communication round at the client side? (in client_app.py)
2. How to implement learning rate scheduling?
Can anyone helps?

**Comment by KarhouTam** [comment]
> Hi, @jafermarq @WilliamLindskog . Feel great to say that the FedBABU reproduction is at the final phase (running experiment with hyperparameters in the paper). Sorry for the very delay about this PR progress, cuz I have other works to do. I have some questions within the reproduction.
> 
> 1. How to get the number of current communication round at the client side? (in client_app.py)
> 2. How to implement learning rate scheduling?
>    Can anyone helps?

This baseline is executable now. But still have some gap with the paper's results, so I'm tunning it now. @WilliamLindskog 
I may need some help for the reproduction.

**Comment by KarhouTam** [comment]
Cannot so far. Need some guidances.

**Comment by jafermarq** [comment]
> Cannot so far. Need some guidances. @jafermarq

For both 1 & 2 you can make use of a call back in your _strategy_. For example, take a look at the `examples/advanced-pytorch`: https://github.com/adap/flower/blob/9c82d166f0fcb9ce908c726d53cbed9de1bf963a/examples/advanced-pytorch/pytorch_example/server_app.py#L36. Note that to use this you don't need a custom strategy necessarily (i.e. you can use vanilla `FedAvg`). The dictionary that the callback returns is what `config` in the client's `fit` method will carry. You can make a small edit to also communicate the server round:

```python
def on_fit_config(server_round: int):
    """Construct `config` that clients receive when running `fit()`"""
    lr = 0.1
    # Enable a simple form of learning rate decay
    if server_round > 10:
        lr /= 2
    return {"lr": lr, "current-round": server_round}
```

Note that this is setting the `config` for the `fit()` method. You can make use of a similar callback but passed as the `on_evaluate_config_fn` arg to the strategy, if you want to also control what the `config` in the client's `evaluate()` method receives.

If you want a more sophisticated LR scheduling you might need to implement it using a custom strategy.

**Comment by KarhouTam** [comment]
Glad to say that the reproduction is finally done. @jafermarq 
This PR is ready for CR.

**Comment by KarhouTam** [comment]
image
the `plot.ipynb` has the results of settings in the screenshot.

Due to my compute resource limit, I only ran 1 settings 😐. @jafermarq

**Comment by KarhouTam** [comment]
Hi,  @jafermarq @WilliamLindskog . I think this PR is  almost completed. Any more changes should I make?
